field/TinyLang/Field/Raw/Core.hs

```diff
@@ -24,16 +26,24 @@ newtype Var = Var { unVar :: Identifier }
     deriving (Eq, Generic)
     deriving (Show) via (Quiet Var)
 
+{-| @Statements v f$, @Statement v f@, and @Expr v f@ are parameterised
+  by the type of variable @v@.
+-}
+newtype Statements v f = Statements { unStatements :: [Statement v f] }
+    deriving (Generic)
+    deriving (Show) via (Quiet (Statements v f))
 
+data Statement v f
+    = ELet    v          (Expr v f)
+    | EAssert (Expr v f)
+    | EFor    v          Integer    Integer (Statements v f)
```

Comment:
Why do we need to wrap `[Statement v f]` with a `newtype` here?

Reply:
TL;DR: so we can reuse the operations that we define over `Statements`.  `[Statements f]` does not exist on its own, therefore I am not keen on it occuring in the AST.

I believe that `Statements f` is semanticaly different than `[Statement f]`. At the moment I expose the accessor, but I do not think that I should be doing it.  With our current AST `[Statement f]` should not be exposed as I think that treating it as a list of statements can lead to potential bugs.

There are some operations that will work on `Statement` and `Program`/`Statements`, but they will not necessarily work on when mapping them over  `[Statement f]`, these operations include:
- renaming, and
- normalisation.

For renaming, we need to get a fresh unique, and that operation cannot be perfomed independently over statements.

For normalisation, we need to have an option of inlining `let` bindings, which you also noticed in a different comment.

Reply:
Ok, please inline these comments somewhere in the code.

I'm not against having a `newtype` wrapper there. I'm against having the same `newtype` wrapper both for the entire program and for the body of a loop. Clearly those are very different things (e.g. to rename a `Program` we need to call `supplyAtLeastFree` over the entire program, but when we rename the body of a loop, there's no such requirement).

So if we can have both `Program = ...` and `Statements = ...`, that will be better, I believe.